localstack/runtime/init.py

```diff
@@ -0,0 +1,223 @@
+
+
+class ScriptRunner:
+    """
+    Interface for running scripts.
+    """
+
+    def run(self, path: str) -> None:
+        """
+        Run the given script with the appropriate runtime.
+
+        :param path: the path to the script
+        """
+        pass
+
+
+class ShellScriptRunner(ScriptRunner):
+    """
+    Runner that interprets scripts as shell scripts and calls them directly.
+    """
+
+    def run(self, path: str) -> None:
+        exit_code = subprocess.call(args=[], executable=path)
+        if exit_code != 0:
+            raise OSError("Script %s returned a non-zero exit code %s" % (path, exit_code))
+
+
+class PythonScriptRunner(ScriptRunner):
+    """
+    Runner that uses ``exec`` to run a python script.
+    """
+
+    def run(self, path: str) -> None:
+        with open(path, "rb") as fd:
+            exec(fd.read())
+
+
+class InitScriptManager:
+    _stage_directories: Dict[Stage, str] = {
+        Stage.BOOT: "boot.d",
+        Stage.START: "start.d",
+        Stage.READY: "ready.d",
+        Stage.SHUTDOWN: "shutdown.d",
+    }
+
+    _script_runners: Dict[str, ScriptRunner] = {
+        ".sh": ShellScriptRunner(),
+        ".py": PythonScriptRunner(),
+    }
+
+    script_root: str
+    stage_completed: Dict[Stage, bool]
+
+    def __init__(self, script_root: str):
+        self.script_root = script_root
+        self.stage_completed = {stage: False for stage in Stage}
+
+    @cached_property
+    def scripts(self) -> Dict[Stage, List[Script]]:
+        return self._find_scripts()
+
+    def get_script_runner(self, script_file: str) -> Optional[ScriptRunner]:
+        for suffix, runner in self._script_runners.items():
+            if script_file.endswith(suffix):
+                return runner
+        return None
+
+    def has_script_runner(self, script_file: str) -> bool:
+        return self.get_script_runner(script_file) is not None
+
+    def run_stage(self, stage: Stage) -> List[Script]:
+        """
+        Runs all scripts in the given stage.
+
+        :param stage: the stage to run
+        :return: the scripts that were in the stage
+        """
+        scripts = self.scripts.get(stage, [])
+
+        if self.stage_completed[stage]:
+            LOG.debug("Stage %s already completed, skipping", stage)
+            return scripts
+
+        try:
+            for script in scripts:
+                LOG.debug("Running %s script %s", script.stage, script.path)
+                try:
+                    script.state = State.RUNNING
+                    runner = self.get_script_runner(script.path)
+                    runner.run(script.path)
+                except Exception as e:
+                    script.state = State.ERROR
+                    if LOG.isEnabledFor(logging.DEBUG):
+                        LOG.exception("Error while running script %s", script)
+                    else:
+                        LOG.exception("Error while running script %s: %s", script, e)
```

Comment:
nit: we could print a warning in the non-DEBUG case (guess that was the original intention?):

```suggestion
                        LOG.warning("Error while running script %s: %s", script, e)
```